### PR TITLE
Carousel: Drag event based index callback

### PR DESCRIPTION
**Pull request by Mitch-At-Work**
## Previous Behavior
Carousel was calling the index change callback on Button and Focus events only, preventing users from receiving callback information about drag events.

## New Behavior
Drag event now correctly emits a native DOM event with type 'drag', this is driven directly by Embla engine and hence is native.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports | Baseline (minified/GZIP) | PR    | Change     |
| :---------------- | -----------------------: | ----: | ---------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ReactComponents.fixture.js'>react-components: entire library</abbr> | `1.159 MB`<br />`290.234 kB` | `1.159 MB`<br />`290.37 kB` | `432 B` <br />`136 B` |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports | Size (minified/GZIP) |
| ----------------- | -------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `69.14 kB`<br />`20.137 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `220.469 kB`<br />`63.821 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `44.447 kB`<br />`14.59 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr> | `8.39 kB`<br />`2.64 kB` |
| <samp>react-timepicker-compat</samp> <br /> <abbr title='bundle-size/TimePicker.fixture.js'>TimePicker</abbr> | `107.387 kB`<br />`35.758 kB` |
</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/ed7c93c5724087f9d2466ee2f4a01b0532d0036d'>ed7c93c5724087f9d2466ee2f4a01b0532d0036d</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">34</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">39</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">81</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Avatar-mount_base.html">631</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Avatar-mount.html">622</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Button-mount_base.html">304</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Button-mount.html">309</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Field-mount_base.html">1103</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Field-mount.html">1128</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProvider-mount_base.html">713</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProvider-mount.html">698</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-mount_base.html">80</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-mount.html">89</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">34</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">39</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">81</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/MakeStyles-mount_base.html">868</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/MakeStyles-mount.html">877</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Persona-mount_base.html">1767</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/Persona-mount.html">1690</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/SpinButton-mount_base.html">1385</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/SpinButton-mount.html">1416</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/SwatchPicker-mount_base.html">1704</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/33092/perf-test-react-components/SwatchPicker-mount.html">1691</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: change/@fluentui-react-carousel-a81a3faa-36db-4ee8-b041-10bc72927e35.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-carousel-a81a3faa-36db-4ee8-b041-10bc72927e35.json:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33092&commitId=8369df67c529887bf7ade38e0a43503c74ed3692&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.Badge Mask RTL.chromium.png|2| Changed|
|Avatar Converged.badgeMask.chromium.png|3| Changed|

</details>


  <details><summary><b>Drawer</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Drawer.overlay drawer full.chromium.png|4| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 2]
```diff
@@ -216,8 +247,9 @@ export function useEmblaCarousel(
   );
 
   React.useEffect(() => {
+    // Scroll to controlled values on update
     const currentActiveIndex = emblaApi.current?.selectedScrollSnap() ?? 0;
-
+    emblaOptions.current.startIndex = activeIndex;
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:252]: The same is done on 262, no?

**Review by layershifter — COMMENTED** [review]
The problem that we have there is kinda simple: Embla does not provide access to events.

IMO, it would nicer to solve it via a custom plugin:

```tsx
export type PointerEventPluginOptions = {
  onSelectViaDrag: (event: PointerEvent | MouseEvent, index: number) => void;
};

export type PointerEventPlugin = CreatePluginType<{}, PointerEventPluginOptions>;

function PointerEventPlugin(options: PointerEventPluginOptions): PointerEventPlugin {
  let emblaApi: EmblaCarouselType;
  let pointerEvent: PointerEvent | MouseEvent | undefined;

  function documentDownListener(event: PointerEvent | MouseEvent) {
    const targetDocument = emblaApi.containerNode().ownerDocument;

    pointerEvent = event;

    targetDocument.removeEventListener('mousedown', documentDownListener);
    targetDocument.removeEventListener('pointerdown', documentDownListener);
  }

  function pointerUpListener() {
    const targetDocument = emblaApi.containerNode().ownerDocument;

    targetDocument.addEventListener('mousedown', documentDownListener);
    targetDocument.addEventListener('pointerdown', documentDownListener);
  }

  function selectListener() {
    if (pointerEvent) {
      const newIndex = emblaApi.selectedScrollSnap() ?? 0;

      options.onSelectViaDrag(pointerEvent, newIndex);
      pointerEvent = undefined;
    }
  }

  function init(emblaApiInstance: EmblaCarouselType, optionsHandler: OptionsHandlerType): void {
    emblaApi = emblaApiInstance;

    emblaApi.on('pointerUp', pointerUpListener);
    emblaApi.on('select', selectListener);
  }

  function destroy(): void {
    emblaApi.off('pointerUp', pointerUpListener);
    emblaApi.off('select', selectListener);
  }

  return {
    name: 'pointerEvent',
    options,
    init,
    destroy,
  };
}
```

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts (hunk 2 above)
Comment (Mitch-At-Work) [reply packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:237]: This one should cover controlled indexes while next is a full reInit.

File: packages/react-components/react-carousel/stories/src/Carousel/CarouselControlled.stories.tsx [hunk 3]
```diff
@@ -128,9 +128,13 @@ export const Controlled = () => {
         activeIndex={activeIndex}
         className={classes.carousel}
         groupSize={1}
-        onActiveIndexChange={(e, data) => setActiveIndex(data.index)}
         announcement={getAnnouncement}
         draggable
+        onActiveIndexChange={(ev, data) => {
+          console.log('Event type:', data.type);
+          console.log('Change:', data.index);
```
Comment (layershifter) [inline packages/react-components/react-carousel/stories/src/Carousel/CarouselControlled.stories.tsx:135]:
```suggestion
```
Oops

File: packages/react-components/react-carousel/stories/src/Carousel/CarouselDefault.stories.tsx [hunk 4]
```diff
@@ -77,7 +77,16 @@ const getAnnouncement: CarouselAnnouncerFunction = (index: number, totalSlides:
...
-  <Carousel groupSize={1} circular announcement={getAnnouncement}>
+  <Carousel
+    draggable
+    groupSize={1}
+    circular
+    announcement={getAnnouncement}
+    onActiveIndexChange={(ev, data) => {
+      console.log('Event type:', data.type);
+      console.log('Change:', data.index);
+    }}
```
Comment (layershifter) [inline packages/react-components/react-carousel/stories/src/Carousel/CarouselDefault.stories.tsx:88]:
```suggestion
```

Oops

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts (hunk 2 above)
Comment (layershifter) [reply packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:237]: Oops, I see. Should we move it inside `if` condition below?

Comment (Mitch-At-Work) [reply packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:237]: I think it makes sense to have these ones segmented as reinit logic should never fire on index change (only enforce retaining the existing index as ref), while we can keep this one specifically local to an index update whether controlled or internal.

File: packages/react-components/react-carousel/stories/src/Carousel/CarouselEventing.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,226 @@
+import {
```
Comment (layershifter) [inline packages/react-components/react-carousel/stories/src/Carousel/CarouselEventing.stories.tsx:1]: Minor UI improvement:

```tsx
import {
  Body1,
  Caption1,
  makeStyles,
  mergeClasses,
  tokens,
  Title1,
  Subtitle2,
  useId,
} from '@fluentui/react-components';
import {
  Carousel,
  CarouselAnnouncerFunction,
  CarouselCard,
  CarouselNav,
  CarouselNavButton,
  CarouselNavContainer,
  CarouselSlider,
  Text,
} from '@fluentui/react-components';
import * as React from 'react';

const useClasses = makeStyles({
  container: {
    display: 'flex',
    gap: '20px',
  },
  carousel: {
    border: `2px solid ${tokens.colorNeutralStroke1}`,
    borderRadius: tokens.borderRadiusMedium,
    boxShadow: tokens.shadow16,
    padding: '20px 0',
    marginTop: '24px'
  },
  card: {
    margin: '10px',
  },
  logLabel: {
    alignSelf: 'end',
    color: tokens.colorNeutralForegroundOnBrand,
    backgroundColor: tokens.colorBrandBackground,
    width: 'fit-content',
    fontWeight: tokens.fontWeightBold,
    padding: '2px 12px',
    borderRadius: `${tokens.borderRadiusMedium} ${tokens.borderRadiusMedium} 0 0`,
  },
  log: {
    overflowY: 'auto',
    boxShadow: tokens.shadow16,
    minWidth: '240px',
    flex: 1,
    border: `2px solid ${tokens.colorBrandBackground}`,
    borderRadius: `${tokens.borderRadiusMedium} 0 ${tokens.borderRadiusMedium} ${tokens.borderRadiusMedium}`,
    padding: '12px',
  },
  logContainer: {
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    filter: `drop-shadow(0 0 4px ${tokens.colorNeutralStroke1})`,
  },
  wireframe: {
    backgroundColor: tokens.colorNeutralBackground3,
    border: `${tokens.strokeWidthThin} solid ${tokens.colorNeutralStroke1}`,

    display: 'flex',
    flexDirection: 'column',
    gap: '12px',
    placeContent: 'center',

    padding: '40px',
    height: '100px',

    position: 'relative',
  },
  wireframeEven: {
    backgroundColor: tokens.colorBrandBackground2,
    border: `${tokens.strokeWidthThin} solid ${tokens.colorBrandStroke1}`,
  },
  wireframeInfo: {
    position: 'absolute',
    right: '12px',
    top: '12px',

    backgroundColor: tokens.colorPaletteRedBackground2,
    border: `${tokens.strokeWidthThin} dotted ${tokens.colorPaletteRedBorder2}`,

    fontSize: tokens.fontSizeBase200,
    padding: '4px 8px',
  },
  wireframeSmall: {
    minWidth: '100px',
    padding: '40px 20px',
  },
  wireframeMedium: {
    minWidth: '200px',
    padding: '40px 20px',
  },
  wireframeLarge: {
    minWidth: '350px',
  },
});

const WireframeContent: React.FC<{
  appearance: 'odd' | 'even';
  children: React.ReactNode;
  size?: 'small' | 'medium' | 'large';
}> = props => {
  const classes = useClasses();

  return (
    <div
      tabIndex={0}
      className={mergeClasses(
        classes.wireframe,
        props.appearance === 'even' && classes.wireframeEven,
        props.size === 'small' && classes.wireframeSmall,
        props.size === 'medium' && classes.wireframeMedium,
        props.size === 'large' && classes.wireframeLarge,
      )}
    >
      <div className={classes.wireframeInfo}>
        <code>size: {props.size ?? 'auto'}</code>
      </div>
      {props.children}
    </div>
  );
};

const getAnnouncement: CarouselAnnouncerFunction = (index: number, totalSlides: number, slideGroupList: number[][]) => {
  return `Carousel slide ${index + 1} of ${totalSlides}`;
};

export const Eventing = () => {
  const classes = useClasses();
  const labelId = useId();

  const [activeIndex, setActiveIndex] = React.useState(0);
  const [statusLog, setStatusLog] = React.useState<
    [number, { type: 'click' | 'focus' | 'drag' | undefined; index: number }][]
  >([]);

  return (
    <div className={classes.container}>
      <Carousel
        className={classes.carousel}
        groupSize={1}
        circular
        draggable
        announcement={getAnnouncement}
        activeIndex={activeIndex}
        onActiveIndexChange={(ev, data) => {
          setActiveIndex(data.index);
          setStatusLog(prev => [[Date.now(), { type: data.type, index: data.index }], ...prev]);
        }}
      >
        <CarouselSlider cardFocus>
          <CarouselCard className={classes.card} autoSize aria-label="1 of 7">
            <WireframeContent appearance="odd">
              <Title1 align="center">Lorem Ipsum</Title1>
              <Body1 align="center">
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor...
              </Body1>
            </WireframeContent>
          </CarouselCard>
          <CarouselCard className={classes.card} autoSize aria-label="2 of 7">
            <WireframeContent appearance="even" size="small">
              <Subtitle2 align="center">Lorem Ipsum</Subtitle2>
              <Caption1 align="center">Lorem ipsum...</Caption1>
            </WireframeContent>
          </CarouselCard>
          <CarouselCard className={classes.card} autoSize aria-label="3 of 7">
            <WireframeContent appearance="odd" size="medium">
              <Title1 align="center">Lorem Ipsum</Title1>
              <Caption1 align="center">Lorem ipsum dolor sit amet...</Caption1>
            </WireframeContent>
          </CarouselCard>
          <CarouselCard className={classes.card} autoSize aria-label="4 of 7">
            <WireframeContent appearance="even" size="large">
              <Title1 align="center">Lorem Ipsum</Title1>
              <Body1 align="center">Lorem ipsum dolor sit amet...</Body1>
            </WireframeContent>
          </CarouselCard>
          <CarouselCard className={classes.card} autoSize aria-label="5 of 7">
            <WireframeContent appearance="odd" size="medium">
              <Title1 align="center">Lorem Ipsum</Title1>
              <Caption1 align="center">Lorem ipsum dolor sit amet...</Caption1>
            </WireframeContent>
          </CarouselCard>
          <CarouselCard className={classes.card} autoSize aria-label="6 of 7">
            <WireframeContent appearance="even" size="large">
              <Title1 align="center">Lorem Ipsum</Title1>
              <Body1 align="center">Lorem ipsum dolor sit amet...</Body1>
            </WireframeContent>
          </CarouselCard>
          <CarouselCard className={classes.card} autoSize aria-label="7 of 7">
            <WireframeContent appearance="odd" size="small">
              <Subtitle2 align="center">Lorem Ipsum</Subtitle2>
              <Caption1 align="center">Lorem ipsum...</Caption1>
            </WireframeContent>
          </CarouselCard>
        </CarouselSlider>

        <CarouselNavContainer
          layout="inline"
          next={{ 'aria-label': 'go to next' }}
          prev={{ 'aria-label': 'go to prev' }}
        >
          <CarouselNav>{index => <CarouselNavButton aria-label={`Carousel Nav Button ${index}`} />}</CarouselNav>
        </CarouselNavContainer>
      </Carousel>

      <div className={classes.logContainer}>
        <div className={classes.logLabel} id={labelId}>
          Events log
        </div>
        <div role="log" aria-labelledby={labelId} className={classes.log}>
          {statusLog.map(([time, status], i) => {
            const date = new Date(time);

            return (
              <div key={i}>
                {date.toLocaleTimeString()}{' '}
                <Text weight="bold">
                  {'{'} type: {status.type}, index: {status.index} {'}'}
                </Text>
              </div>
            );
          })}
        </div>
      </div>
    </div>
  );
};

Eventing.parameters = {
  docs: {
    description: {
      story: 'Carousel provides callbacks on index change with a multitude of event types.',
    },
  },
};
```

**Review by layershifter — APPROVED** [review]

**Pull request merged**